test/test_selectors.py

```diff
@@ -0,0 +1,684 @@
+        if self.lower is not None:
+            self.testcase.assertGreaterEqual(total_time, self.lower - TOLERANCE)
+        if self.upper is not None:
+            self.testcase.assertLessEqual(total_time, self.upper + TOLERANCE)
+
+
+class TimerMixin(object):
+    def assertTakesTime(self, lower=None, upper=None):
+        return TimerContext(self, lower=lower, upper=upper)
+
+
+@skipUnless(selectors.HAS_SELECT, "Platform doesn't have a selector")
+class WaitForIOTest(unittest.TestCase, AlarmMixin, TimerMixin):
+    """ Tests for the higher level wait_for_* functions. """
+    def make_socketpair(self):
+        rd, wr = socketpair()
+        self.addCleanup(rd.close)
+        self.addCleanup(wr.close)
+        return rd, wr
+
+    def test_selector_error(self):
+        err = selectors.SelectorError(1)
+        self.assertEqual(err.__repr__(), "<SelectorError errno=1>")
+        self.assertEqual(err.__str__(), "<SelectorError errno=1>")
+
+    def test_wait_for_read_single_socket(self):
+        rd, wr = self.make_socketpair()
+        self.assertEqual([], selectors.wait_for_read(rd, timeout=0.001))
+
+    def test_wait_for_read_multiple_socket(self):
+        rd, rd2 = self.make_socketpair()
+        self.assertEqual([], selectors.wait_for_read([rd, rd2], timeout=0.001))
+
+    def test_wait_for_read_empty(self):
+        self.assertEqual([], selectors.wait_for_read([], timeout=0.001))
+
+    def test_wait_for_write_single_socket(self):
+        wr, wr2 = self.make_socketpair()
+        self.assertEqual([wr], selectors.wait_for_write(wr, timeout=0.001))
+
+    def test_wait_for_write_multiple_socket(self):
+        wr, wr2 = self.make_socketpair()
+        result = selectors.wait_for_write([wr, wr2], timeout=0.001)
+        # assertItemsEqual renamed in Python 3.x
+        if hasattr(self, "assertItemsEqual"):
+            self.assertItemsEqual([wr, wr2], result)
+        else:
+            self.assertCountEqual([wr, wr2], result)
+
+    def test_wait_for_write_empty(self):
+        self.assertEqual([], selectors.wait_for_write([], timeout=0.001))
+
+    def test_wait_for_non_list_iterable(self):
+        rd, wr = self.make_socketpair()
+        iterable = {'rd': rd}.values()
+        self.assertEqual([], selectors.wait_for_read(iterable, timeout=0.001))
+
+    def test_wait_timeout(self):
+        rd, wr = self.make_socketpair()
+        with self.assertTakesTime(lower=0.0, upper=0.1):
+            selectors.wait_for_read([rd], timeout=0.01)
+
+    @skipUnless(hasattr(signal, "alarm"), "Platform doesn't have signal.alarm()")
+    def test_interrupt_wait_for_read_no_event(self):
+        rd, wr = self.make_socketpair()
+
+        sigalrm_handler = signal.signal(signal.SIGALRM, lambda *args: None)
+        self.addCleanup(signal.signal, signal.SIGALRM, sigalrm_handler)
+
+        # Start the timer for the interrupt.
+        self.set_alarm(SHORT_SELECT)
+
+        with self.assertTakesTime(lower=LONG_SELECT):
```

Comment:
Basically we want our tests to be as accurate as possible. While random failures on the CI are possible, it just means our tolerances aren't high enough. :grin: